src/lib/y2storage/package_handler.rb

```diff
@@ -116,6 +116,21 @@ def set_proposal_packages
       success
     end
 
+    # Add the proposal packages for storage that are needed for the specified
+    # devicegraph's used features. This marks the packages for installation;
+    # it does not install them yet.
+    #
+    # @param devicegraph [Devicegraph] usually StorageManager.instance.staging
+    # @param optional
+    def self.set_proposal_packages_for(devicegraph, optional: true)
```

Comment:
This is literally just moving that piece of code from the _inst_disk_proposal_ client to another place to be able to share the code.

Since the two clients that need that method, _inst_disk_proposal_ and now also _partitions_proposal_ (used exclusively by MicroOS) are structured so differently, it would be very awkward to make both of them inherit a new base class just for this purpose; that would not be a good use of inheritance. Refactoring both of them to share more than just that one method would involve considerable risk of breaking things in released products. So it would really be a base class with just that one method that both clients would share; that feels just wrong.

I thought long about where to put this method, but there was no better place. Creating a one-method class for this also feels very wrong; and the obvious name for such a class would be something along the lines of `PackageHandler` (`PackageManager` is in use in the YaST/ZyPP universe already).

But we already _have_ a class named `PackageHandler`; which evolved over time due to ever-changing requirements and quirks of certain products, so the original design doesn't fit exactly anymore. The idea of discriminating between required and optional packages had crept in over time, and once the class was adapted to that, it was never intended to handle _both_ at the same time, required and optional packages.

That shortcoming had been crudely papered over by that old `add_storage_packages()` method in the _inst_disk_proposal_ client by simply creating _two_ instances of `PackageHandler`, one for the required packages and one for the optional ones.

The need to do that goes back even further to a time when the used features in libstorage were designed to only handle one kind of them at a time, either all used features or only the required used features.

There is a lot of history and old design choices that were contradicted by those ever-changing requirements (which of course were never specified or recorded or collected anywhere) going on.


Maybe some day we might want to refactor all that. But today is not that day. We have a very concrete bug that needs to be fixed (perferably before somebody elevates it to a security problem), and this - admittedly crude and not very elegant - change is a working for for it.

Reply:
JFTR: We both agreed on a Jangouts call that this is acceptable as a pragmatic approach.